gradle/generation/snowball.gradle

```diff
@@ -0,0 +1,118 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+apply plugin: "de.undercouch.download"
+
+configure(rootProject) {
+  ext {
+    snowballWorkDir    = file("${buildDir}/snowball")
+    snowballStemmerDir = file("${buildDir}/snowball/stemmers")
+    snowballPatchedDir = file("${buildDir}/snowball/patched")
+    snowballWebsiteDir = file("${buildDir}/snowball/website")
+    snowballDataDir    = file("${buildDir}/snowball/data")
+  }
+
+  task snowball()  {
+    description "Regenerate snowball-based sources, stopwords, and tests for ...lucene/analysis."
+    group "generation"
+
+    dependsOn ":lucene:analysis:common:snowballGen"
+  }
+
+  task downloadSnowballStemmers(type: Download) {
+    def stemmerZip = file("${snowballWorkDir}/stemmers.zip")
+
+    src "https://github.com/snowballstem/snowball/archive/53739a805cfa6c77ff8496dc711dc1c106d987c1.zip"
+    dest stemmerZip
+    onlyIfModified true
+    useETag "all"
+    cachedETagsFile file("${snowballWorkDir}/stemmers.json")
+
+    doLast {
+      ant.unzip(src: stemmerZip, dest: snowballStemmerDir, overwrite: "true") {
+        ant.cutdirsmapper(dirs: "1")
+      }
+    }
+  }
+
+  task downloadSnowballWebsite(type: Download) {
+    def websiteZip = file("${snowballWorkDir}/website.zip")
+
+    src "https://github.com/snowballstem/snowball-website/archive/ff891e74f08e7315523ee3c0cad55bb1b7831b9d.zip"
+    dest websiteZip
+    onlyIfModified true
+    useETag "all"
+    cachedETagsFile file("${snowballWorkDir}/website.json")
+
+    doLast {
+      ant.unzip(src: websiteZip, dest: snowballWebsiteDir, overwrite: "true") {
+        ant.cutdirsmapper(dirs: "1")
+      }
+    }
+  }
+
+  task downloadSnowballData(type: Download) {
+    def dataZip = file("${snowballWorkDir}/data.zip")
+
+    src "https://github.com/snowballstem/snowball-data/archive/9145f8732ec952c8a3d1066be251da198a8bc792.zip"
+    dest dataZip
+    onlyIfModified true
+    useETag "all"
+    cachedETagsFile file("${snowballWorkDir}/data.json")
+
+    doLast {
+      ant.unzip(src: dataZip, dest: snowballDataDir, overwrite: "true") {
+        ant.cutdirsmapper(dirs: "1")
+      }
+    }
+  }
+
+  task cleanSnowballCheckout(type: Delete) {
+    dependsOn downloadSnowballStemmers
+    delete snowballPatchedDir
+  }
+
+  task patchSnowball(type: Copy) {
+    dependsOn cleanSnowballCheckout
+
+    from fileTree(snowballStemmerDir) {
+      include '**/*'
+    }
+    into snowballPatchedDir
+
+    doLast {
+      ant.patch(patchfile: rootProject.file("gradle/generation/snowball.patch"), dir: snowballPatchedDir, strip: "1")
+    }
+  }
+}
+
+configure(project(":lucene:analysis:common")) {
+  task snowballGen() {
+    description "Patch and Regenerate snowball sources, stopwords, and tests"
+    group "generation"
+    dependsOn rootProject.patchSnowball
```

Comment:
Why not move all these tasks into lucene:analysis:common configuration? Are they reused anywhere else? If not then it'd make everything local to the project.

Reply:
I can look into it. I am pretty clueless on gradle, I based this on the `moman` regeneration that is located in `gradle/generation/util.gradle`. Because it is doing similar stuff (downloading zip and running commands).

Reply:
Commit it in, I'll correct it afterwards, no worries.

Reply:
ok I think this is now better. Now tasks are under analysis common, downloads stuff to `lucene/analysis/common/build` instead of `build/` and all tasks are under that project, except a single top-level one with the description that depends on it.

Reply:
Thanks Robert. FYI: the moman build could be modified the same way. There is a root project dependency section in javacc.gradle but the rationale behind it is that it's a configuration dependency then shared by different subprojects (and a task that requires javacc).